### PR TITLE
NIFI-5481 Additional Sensitive Property Providers

### PR DESCRIPTION
**Pull request by natural**
### Description of PR 

The code in this change-set provides the following:

  - refactors Sensitive Property Providers
  - adds AWS KMS sensitive property provider (https://issues.apache.org/jira/browse/NIFI-6325)
  - adds Key Store sensitive property provider (https://issues.apache.org/jira/browse/NIFI-6326)
  - adds HashiCorp Vault sensitive property provider (https://issues.apache.org/jira/browse/NIFI-6324)
  - adds GCP sensitive property provider (https://issues.apache.org/jira/browse/NIFI-6616)
  - adds Azure sensitive property provider (https://issues.apache.org/jira/browse/NIFI-6515)
  - integrates the above with the Config Encryption Toolkit (https://issues.apache.org/jira/browse/NIFI-6363)
  - unit tests, integration tests, documentation for the above

This PR replaces #3542.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-docs/src/main/asciidoc/administration-guide.adoc [hunk 1]
```diff
@@ -1720,9 +1720,9 @@ For each Node, the minimum properties to configure are as follows:
    thread pool will never have fewer than this number of threads. It will grow as needed up to the maximum value set by the `nifi.cluster.node.protocol.max.threads`
    property.
 ** `nifi.cluster.node.protocol.max.threads` - The maximum number of threads that should be used to communicate with other nodes in the cluster. This property
-	defaults to `50`. A thread pool is used for replication requests to all nodes, and the thread pool will have a "core" size that is configured by the
```
Comment (pvillard31) [inline nifi-docs/src/main/asciidoc/administration-guide.adoc]: Could you revert all the unnecessary changes on the documentation?

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    private String locationId;
+    private String keyRingId;
+    private String cryptoKeyId;
+
+    private final String resource;
+    private final KeyManagementServiceClient client;
+
+
+    public GCPKMSSensitivePropertyProvider(String keyId) {
+        if (StringUtils.isBlank(System.getenv("GOOGLE_APPLICATION_CREDENTIALS"))) {
```
Comment (pvillard31) [inline nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java]: Given the nature of NiFi, I'd try, whenever possible, to avoid using environment variables. IIRC it should be possible to initialize the KMS client using ``KeyManagementServiceSettings`` and provide a path to the service account JSON file. The reason I'd recommend that is to avoid exposing the environment variable to users through the expression language (and give the users the associated service account permissions). However it'd require adding a NiFi property to configure that path. Thoughts? (cc @alopresto to be sure I'm not saying anything stupid)

File: nifi-commons/nifi-security-utils/src/main/java/org/apache/nifi/security/util/CipherUtils.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+import java.security.SecureRandom;
+import java.security.Security;
+
+
+/**
+ * Common functionality for ciphers:  pre-initialized ciphers, IVs, random value generators, etc.
+ */
+public class CipherUtils {
+    final static SecureRandom random = new SecureRandom();
+    public final static int IV_LENGTH = 12;
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-commons/nifi-security-utils/src/main/java/org/apache/nifi/security/util/CipherUtils.java]: Good catch, added.

File: nifi-commons/nifi-security-utils/src/test/java/org/apache/nifi/security/util/CipherUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        byte [][] ivs = new byte[count][];
+        for (int i = 0; i < count; i++) {
+            ivs[i] = CipherUtils.generateIV();
+        }
+
+        // Compare each IV to every other IV
+        for (int i = 0; i < count; i++) {
+            byte[] iv = ivs[i];
+
+            for (int j = 0; j < count; j++) {
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-commons/nifi-security-utils/src/test/java/org/apache/nifi/security/util/CipherUtilsTest.java]: Great example, note added.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/ExternalProperties.java [hunk 5]
```diff
@@ -14,10 +14,21 @@
...
-    SensitivePropertyProvider getProvider();
+    /**
+     * Read an external property by name.
+     *
+     * @param name the name or key of the external property
+     * @return external property value, as a string
+     */
+    String get(String name);
 
+    String get(String name, String missing);
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/ExternalProperties.java]: Added, thanks.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/ProtectedNiFiProperties.java [hunk 6]
```diff
@@ -72,10 +75,19 @@ public ProtectedNiFiProperties(NiFiProperties props) {
      *
      * @param rawProps the Properties to contain
      */
-    public ProtectedNiFiProperties(Properties rawProps) {
-        this(new StandardNiFiProperties(rawProps));
+    public ProtectedNiFiProperties(Properties rawProps, SensitivePropertyProvider sensitivePropertyProvider) {
+        this(new StandardNiFiProperties(rawProps), sensitivePropertyProvider);
     }
 
+    public ProtectedNiFiProperties(NiFiProperties props, String keyOrKeyId) {
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/ProtectedNiFiProperties.java]: Added.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/SensitivePropertyConfigurationException.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.properties.sensitive;
+
+public class SensitivePropertyConfigurationException extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/SensitivePropertyConfigurationException.java]: Thanks for the suggestion.  Added exception hierarchy and reviewed for simplification opportunities.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/StandardExternalPropertyLookup.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * Slurps property values from the usual places in the usual order.
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/StandardExternalPropertyLookup.java]: Good catch, updated comments.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/aes/AESSensitivePropertyProvider.java [hunk 9]
```diff
@@ -138,7 +146,7 @@ public String getName() {
      */
     @Override
     public String getIdentifierKey() {
-        return IMPLEMENTATION_KEY + getKeySize(Hex.toHexString(key.getEncoded()));
+        return IMPLEMENTATION_KEY + Hex.toHexString(key.getEncoded()); // getKeySize();
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/aes/AESSensitivePropertyProvider.java]: Oof, yeah, that wasn't supposed to end up this way.  I've reverted all of the bits related to this.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/azure/keyvault/AzureKeyVaultSensitivePropertyProvider.java [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+    public AzureKeyVaultSensitivePropertyProvider(String material) {
+        final File credFile = new File(System.getenv("AZURE_AUTH_LOCATION"));
+        try {
+            ApplicationTokenCredentials.fromFile(credFile).clientId();
+        } catch (IOException e) {
+            throw new SensitivePropertyConfigurationException(e);
+        }
+
+        String prefix = MATERIAL_PREFIX + "/" + MATERIAL_KEY_TYPE + "/";
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/azure/keyvault/AzureKeyVaultSensitivePropertyProvider.java]: Fixed.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java (hunk 2 above)
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java]: Thanks for the feedback.  I'll double back on this and change the auth if I can.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+        extractKeyParts(keyId);
+        this.resource = CryptoKeyName.format(projectId, locationId, keyRingId, cryptoKeyId);
+        try {
+            this.client = KeyManagementServiceClient.create();
+        } catch (IOException e) {
+            throw new SensitivePropertyConfigurationException("Unable to create service client", e);
+        }
+    }
+
+    private void extractKeyParts(String keyId) {
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java]: Renamed to `setKeyParts` to better reflect intention + added javadocs.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+        if (StringUtils.isBlank(keyId)) throw new SensitivePropertyConfigurationException("The key cannot be empty");
+
+        this.keyId = keyId;
+
+        String[] parts = this.keyId.split(MATERIAL_SEPARATOR, 3);
+        if (parts.length != 3 || !parts[0].equals(MATERIAL_PREFIX) || !parts[1].equals(MATERIAL_KEY_TYPE))
+            throw new SensitivePropertyConfigurationException("Invalid GCP key");
+
+        String path = parts[2];
+        Pattern simplePattern = Pattern.compile("([^/]+)/([^/]+)/([^/]+)/([^/]+)");
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java]: Migrated, thanks.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+
+        if (match == null || match.groupCount() != 4 || !match.find())
+            throw new SensitivePropertyConfigurationException("Invalid GCP key pattern");
+
+        projectId = match.group(1);
+        locationId = match.group(2);
+        keyRingId = match.group(3);
+        cryptoKeyId = match.group(4);
+
+        if (StringUtils.isBlank(projectId) || StringUtils.isBlank(locationId) || StringUtils.isBlank(keyRingId) || StringUtils.isBlank(cryptoKeyId))
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java]: Great catch, updated.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+    @Override
+    public String protect(String unprotectedValue) throws SensitivePropertyProtectionException {
+        if (StringUtils.isBlank(unprotectedValue)) {
+            throw new IllegalArgumentException("Cannot encrypt an empty value");
+        }
+
+        EncryptResponse response;
+        try {
+            response = client.encrypt(resource, ByteString.copyFrom(unprotectedValue, StandardCharsets.UTF_8));
+        } catch (final com.google.api.gax.rpc.NotFoundException e) {
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java]: Thanks for catching that, added imports.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+     */
+    @Override
+    public String unprotect(String protectedValue) throws SensitivePropertyProtectionException {
+        if (StringUtils.isBlank(protectedValue)) {
+            throw new IllegalArgumentException("Cannot decrypt empty or blank cipher text.");
+        }
+        DecryptResponse response;
+        try {
+            response = client.decrypt(resource, ByteString.copyFrom(Base64.decode(protectedValue)));
+        } catch (final org.bouncycastle.util.encoders.DecoderException | com.google.api.gax.rpc.InvalidArgumentException | com.google.api.gax.rpc.NotFoundException e) {
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java]: Imported + updated.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/keystore/ByteArrayKeyStoreProvider.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
+/*
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/keystore/ByteArrayKeyStoreProvider.java]: Moved.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/keystore/KeyStoreSensitivePropertyProvider.java [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Sensitive properties using KeyStore keys with an inner AES SPP.
+ */
+public class KeyStoreSensitivePropertyProvider implements SensitivePropertyProvider {
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/keystore/KeyStoreSensitivePropertyProvider.java]: Renamed to match intent.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/test/groovy/org/apache/nifi/properties/sensitive/aes/AESSensitivePropertyProviderTest.groovy [hunk 18]
```diff
@@ -381,6 +389,7 @@ class AESSensitivePropertyProviderTest extends GroovyTestCase {
         }
     }
 
+    @Ignore // for now bc size is gone
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/test/groovy/org/apache/nifi/properties/sensitive/aes/AESSensitivePropertyProviderTest.groovy]: Related to the erroneous change to the `aes/gcm/{suffix}` format, reverted.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/StandardSensitivePropertyProvider.java [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+     * @param key provider encryption key
+     * @param scheme name of encryption or protection scheme
+     * @return concrete instance of SensitivePropertyProvider, or null when no key/key id is specified
+     * @throws SensitivePropertyProtectionException when a key/key id is not handled by any provider.
+     */
+    public static SensitivePropertyProvider fromKey(String key, String scheme) {
+        if (StringUtils.isBlank(key)) {
+            return null;
+
+            // TODO:  remove these scheme checks
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/StandardSensitivePropertyProvider.java]: Cleaned up; thanks.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/test/java/org/apache/nifi/properties/sensitive/hashicorp/vault/VaultHttpSensitivePropertyProviderIT.java [hunk 20]
```diff
@@ -0,0 +1,421 @@
...
+        operations.write("cubbyhole/token", config);
+        logger.info("Constructed 'cubbyhole token' with token-id:" + cubbyholeTokenId);
+    }
+
+    @AfterClass
+    public static void stopServerContainer() {
+        try {
+            vaultContainer.stop();
+        } catch (final Exception ignored) {
+        }
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/test/java/org/apache/nifi/properties/sensitive/hashicorp/vault/VaultHttpSensitivePropertyProviderIT.java]: Added a `finally {}` that retries the `.stop()`.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-runtime/src/main/java/org/apache/nifi/NiFi.java [hunk 21]
```diff
@@ -353,14 +353,13 @@ private static String loadFormattedKey(String[] args) {
         if (parsedArgs.contains(KEY_FILE_FLAG)) {
             key = getKeyFromKeyFileAndPrune(parsedArgs);
             // Format the key (check hex validity and remove spaces)
-            key = formatHexKey(key);
-
+            // key = formatHexKey(key);
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-runtime/src/main/java/org/apache/nifi/NiFi.java]: Removed, thanks for catching that.

File: nifi-docs/src/main/asciidoc/administration-guide.adoc (hunk 1 above)
Comment (natural) [reply nifi-docs/src/main/asciidoc/administration-guide.adoc]: Oof, yeah.  I would blame my tools if I could.  Thanks for the catch.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/test/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProviderIT.java [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+            keyId = "gcp/kms/" + createdKey.getName();
+            logger.info("Created GCP key: {}", createdKey.getName());
+
+        } catch (IOException e) {
+            logger.warn("Error during create key ring and key: " + e);
+        }
+    }
+
+    @AfterClass
+    public static void deleteKeyRingAndKey() {
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/test/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProviderIT.java]: Reviewed all 6 provider test cases for this possibility, found none, and commented on each.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/test/java/org/apache/nifi/properties/sensitive/aws/kms/AWSKMSSensitivePropertyProviderIT.java [hunk 23]
```diff
@@ -0,0 +1,220 @@
...
+
+    private static String[] knownGoodKeys;
+    private static AWSKMSClient client;
+
+    /**
+     * Before the tests are run, this method reads the aws credentials file, and when successful, sets those values as
+     * system properties.
+     */
+    @BeforeClass
+    public static void setUpOnce() throws Exception {
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/test/java/org/apache/nifi/properties/sensitive/aws/kms/AWSKMSSensitivePropertyProviderIT.java]: Added comments to all 6 provider test cases explaining the setup necessary.  For those that require specific user values (AWS, Azure, GCP), we have checks to ensure the tests only run when that setup is present.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/test/java/org/apache/nifi/properties/sensitive/azure/keyvault/AzureKeyVaultSensitivePropertyProviderIT.java [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+import org.apache.nifi.properties.sensitive.SensitivePropertyProvider;
+import org.junit.Assume;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class AzureKeyVaultSensitivePropertyProviderIT extends AbstractSensitivePropertyProviderTest {
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/test/java/org/apache/nifi/properties/sensitive/azure/keyvault/AzureKeyVaultSensitivePropertyProviderIT.java]: Added.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/hashicorp/vault/VaultSensitivePropertyProvider.java [hunk 25]
```diff
@@ -0,0 +1,293 @@
...
+            throw new SensitivePropertyConfigurationException("The key cannot be empty");
+
+        VaultEndpoint vaultEndpoint = VaultEndpoint.from(URI.create(serverUri));
+        StandardVaultConfiguration config = new StandardVaultConfiguration(vaultEndpoint, externalProperties);
+
+        Resource keyStore = config.getVaultSslKeyStore();
+        Resource trustStore = config.getVaultSslTrustStore();
+        SslConfiguration sslConf = SslConfiguration.NONE;
+
+        if (keyStore != null || trustStore != null) {
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/hashicorp/vault/VaultSensitivePropertyProvider.java]: As far as I can see, the issue is that `SslConfiguration` requires both stores to be non-`null`.  I've re-factored the above block to be a bit more explicit + require both config types, and I've added integration tests around the behavior.

(We could drop key store completely at the moment, but I didn't want to do that in case we ever need to implement client cert authentication)

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java (hunk 2 above)
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/gcp/kms/GCPKMSSensitivePropertyProvider.java]: I'm reluctant to change this behavior because it matches the recommendations (https://cloud.google.com/docs/authentication/getting-started).

Also, these SPP classes need their configuration to exist outside of nifi (in system props, environment vars, etc) so I'm not sure if we can prevent bad things from happening in these cases.

For now I think we're going to resolve this via documentation.

Thanks again for the feedback.  If I've missed something or if you feel strongly about this, please let me know.

File: nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/azure/keyvault/AzureKeyVaultSensitivePropertyProvider.java [hunk 26]
```diff
@@ -0,0 +1,162 @@
...
+    /**
+     * Returns the "protected" form of this value. This is a form which can safely be persisted in the {@code nifi.properties} file without compromising the value.
+     * An encryption-based provider would return a cipher text, while a remote-lookup provider could return a unique ID to retrieve the secured value.
+     *
+     * @param unprotectedValue the sensitive value
+     * @return the value to persist in the {@code nifi.properties} file
+     */
+    @Override
+    public String protect(String unprotectedValue) throws SensitivePropertyProtectionException {
+        // TODO:  check length of value; azure limits length of encryption operations.
```
[the remark this answers is not in the document]
Comment (natural) [reply nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-properties-loader/src/main/java/org/apache/nifi/properties/sensitive/azure/keyvault/AzureKeyVaultSensitivePropertyProvider.java]: The MS docs do not state the max size, but instead defer to the RFCs for the various JWT algos.  Suspicious, I instead wrote a little unit test to discover the maximum length.  After I did that, I added the constant as a field and a check in the call to `protect`

**Comment by natural** [comment]
Resubmitting in #3916.
